File: server/mbservice/src/main/java/org/rliz/mbs/release/repository/ReleaseGroupRepository.java
package org.rliz.mbs.release.repository;

import org.rliz.mbs.release.model.ReleaseGroup;
import org.springframework.data.jpa.repository.EntityGraph;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;
import java.util.UUID;

/**
 * Repository for {@link ReleaseGroup}s.
 */
public interface ReleaseGroupRepository extends JpaRepository<ReleaseGroup, Long> {

    /**
     * Retrieve single {@link ReleaseGroup} by its UUID.
     *
     * @param identifier the UUID
     * @return the {@link ReleaseGroup}
     */
    @EntityGraph(attributePaths = "artistCredit.artistCreditName.artist")
    ReleaseGroup findByIdentifier(UUID identifier);

    /**
     * Retrieve all {@link ReleaseGroup}s for a given list of artists.
     *
     * @param artistNames the names of the credited artists
     * @return the list of found {@link ReleaseGroup}s
     */
    @EntityGraph(attributePaths = {"artistCredit.artistCreditName.artist"})
    @Query("SELECT r FROM ReleaseGroup r join r.artistCredit ac join ac.artistCreditName acn join acn.artist a " +
            "where lower(a.name) in (:artistNames)")
    List<ReleaseGroup> findReleaseGroupByArtistNames(@Param("artistNames") List<String> artistNames);

}

File: server/mbservice/src/main/java/org/rliz/mbs/release/repository/ReleaseGroupRepository.kt
package org.rliz.mbs.release.repository

import org.rliz.mbs.release.model.ReleaseGroup
import org.springframework.data.jpa.repository.EntityGraph
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import java.util.*

interface ReleaseGroupRepository : JpaRepository<ReleaseGroup, Long> {

    @EntityGraph(attributePaths = arrayOf("artistCredit.artistCreditName.artist"))
    fun findByIdentifier(identifier: UUID): ReleaseGroup

    @EntityGraph(attributePaths = arrayOf("artistCredit.artistCreditName.artist"))
    fun findByIdentifierIn(identifiers: List<UUID>): List<ReleaseGroup>

    @Query("""
        SELECT r FROM
        ReleaseGroup r
        join r.artistCredit ac
        join ac.artistCreditName acn
        join acn.artist a
        where lower(a.name) in (:artistNames)
        """)
    @EntityGraph(attributePaths = arrayOf("artistCredit.artistCreditName.artist"))
    fun findReleaseGroupByArtistNames(@Param("artistNames") artistNames: List<String>): List<ReleaseGroup>

}
